### ISSUE TITLE
The action channel has perceived duplicated action traces for transfers

### ISSUE THREAD

**Issue by coachbjork** [eosswedenorg/thalos#29]
The action channel has one message for each trace in the action, resulting in multiple messages from Thalos to the client from one action. There seem to be one message for each notified account.

``` 
{"tx_id":"2efa5c1ce9ad56badd46fc95e15287508b839762a3d6e9c2709d558f1c45c9bd","blocknum":285784703,"blocktimestamp":"2024-01-04T10:26:02.500","receipt":{"receiver":"eosio.token","act_digest":"5842caf248aa6bb205db9a0c45eeed2811dfb4b167e78590eed333349b2b29b5","global_sequence":87600859834,"recv_sequence":87600859834,"auth_sequence":[{"account":"test2.dp","sequence":1064317}],"code_sequence":2,"abi_sequence":2},"name":"transfer","contract":"eosio.token","receiver":"eosio.token","data":{"from":"test2.dp","memo":"Pong","quantity":"0.00010000 WAX","to":"test1.dp"},"authorization":[{"actor":"test2.dp","permission":"active"}],"except":"","error":0,"return":null}

{"tx_id":"2efa5c1ce9ad56badd46fc95e15287508b839762a3d6e9c2709d558f1c45c9bd","blocknum":285784703,"blocktimestamp":"2024-01-04T10:26:02.500","receipt":{"receiver":"test2.dp","act_digest":"5842caf248aa6bb205db9a0c45eeed2811dfb4b167e78590eed333349b2b29b5","global_sequence":87600859835,"recv_sequence":87600859835,"auth_sequence":[{"account":"test2.dp","sequence":1064318}],"code_sequence":2,"abi_sequence":2},"name":"transfer","contract":"eosio.token","receiver":"test2.dp","data":{"from":"test2.dp","memo":"Pong","quantity":"0.00010000 WAX","to":"test1.dp"},"authorization":[{"actor":"test2.dp","permission":"active"}],"except":"","error":0,"return":null}

{"tx_id":"2efa5c1ce9ad56badd46fc95e15287508b839762a3d6e9c2709d558f1c45c9bd","blocknum":285784703,"blocktimestamp":"2024-01-04T10:26:02.500","receipt":{"receiver":"test1.dp","act_digest":"5842caf248aa6bb205db9a0c45eeed2811dfb4b167e78590eed333349b2b29b5","global_sequence":87600859836,"recv_sequence":87600859836,"auth_sequence":[{"account":"test2.dp","sequence":1064319}],"code_sequence":2,"abi_sequence":2},"name":"transfer","contract":"eosio.token","receiver":"test1.dp","data":{"from":"test2.dp","memo":"Pong","quantity":"0.00010000 WAX","to":"test1.dp"},"authorization":[{"actor":"test2.dp","permission":"active"}],"except":"","error":0,"return":null} 
```

All 3 of these messages are in fact the same action. 
But for the client it could duplicate the information depending on how it's processed.

After discussion with @pnx, a cache in Thalos could solve this for the action channel.

Not sure exactly which information that should be cached for Thalos to identify each action.

The act_digest is unique the the action data, but not to the action itself. transfer from account1 to account2, for the same amount and same memo, will have the same hash. 

Each transfer has the receiver:contract + act_digest, which is a unique combination per action. Which is the first message sent out in a transfer.

In contracts, you look at the first_receiver to ensure it's a unique action (Thanks Mike for teaching me this), to prevent double spending. Is this something that can be applied here as well?

Does the traces have more info than what is forwarded to Thalos? Is there some unique identifier per action there already that we are not currently looking at?

**Comment by pnx** [comment]
Just to clarify. I think the word "cache" is here is a bit misleading, as a cache will typically have longer lifetime than the lifetime of a function call. What I am thinking is more like a lookup table that only exists for one transaction.

Some pseudocode below:

```js
// Called for each transaction
onTransaction(transaction) {
    var lookup_table;
    foreach(transaction.actions as action) {
        if (lookup_table.has(action)) {
            continue;
        }
        lookup_table.insert(action);

        // Process
    }
}
````

I know that we used cache when discussing and it is not wrong, just different lifetimes and I don't think we need to store the information for that long.

**Comment by pnx** [comment]
Fix implemented in v1.1.0

There is now a flag `first_receiver` in an action trace message that tells the user if the action is for the first receiver.

The implementation is simple. `first_receiver` is true if  the `receiver` field is equal to the `contract` field. a contract action must atleast notify itself, so there is always this type of action. I don't know if there is a possibility for a contract to notify itself more than once (and if it is, maybe those contracts should not do that 🤣)
